Pyrado/pyrado/algorithms/meta/pddr.py

```diff
@@ -0,0 +1,328 @@
+                )
+                self.prune_teachers()
+            assert self.num_teachers == len(self.teacher_policies)
+            self.reset_checkpoint()
+
+        # Student
+        self._expl_strat = NormalActNoiseExplStrat(self._policy, std_init=std_init)
+        self._policy = self._policy.to(self.device)
+        self.optimizer = to.optim.Adam([{"params": self.policy.parameters()}], lr=lr)
+
+        # Environments
+        self.samplers = [
+            ParallelRolloutSampler(
+                self.teacher_envs[t],
+                deepcopy(self._expl_strat),
+                num_workers=int(self.num_cpu / self.num_teachers),
+                min_steps=self.min_steps,
+            )
+            for t in range(self.num_teachers)
+        ]
+
+        self.teacher_weights = np.ones(self.num_teachers)
+
+        # Distillation loss criterion
+        self.criterion = to.nn.KLDivLoss(log_target=True, reduction="batchmean")
+
+    @property
+    def expl_strat(self) -> NormalActNoiseExplStrat:
+        return self._expl_strat
+
+    def step(self, snapshot_mode: str, meta_info: dict = None):
+        """
+        Perform a single iteration of the algorithm. This includes collecting the data, updating the parameters, and
+        adding the metrics of interest to the logger. Does not update the `curr_iter` attribute.
+
+        :param snapshot_mode: determines when the snapshots are stored (e.g. on every iteration or on new highscore)
+        :param meta_info: is not `None` if this algorithm is run as a subroutine of a meta-algorithm,
+                          contains a dict of information about the current iteration of the meta-algorithm
+        """
+        # Save snapshot to save the correct iteration count
+        self.save_snapshot()
+
+        if self.curr_checkpoint == -1:
+            self.train_teachers(snapshot_mode, None)
+            self.reached_checkpoint()  # setting counter to 0
+
+        if self.curr_checkpoint == 0:
+            # Sample observations
+            ros, rets, all_lengths = self.sample()
+
+            # Log current progress
+            self.logger.add_value("max return", np.max(rets), 4)
+            self.logger.add_value("median return", np.median(rets), 4)
+            self.logger.add_value("avg return", np.mean(rets), 4)
+            self.logger.add_value("min return", np.min(rets), 4)
+            self.logger.add_value("std return", np.std(rets), 4)
+            self.logger.add_value("std var", self.expl_strat.std.item(), 4)
+            self.logger.add_value("avg rollout len", np.mean(all_lengths), 4)
+            self.logger.add_value("num total samples", np.sum(all_lengths))
+
+            # Save snapshot data
+            self.make_snapshot(snapshot_mode, np.mean(rets), meta_info)
+
+            # Update policy and value function
+            self.update(rollouts=ros)
+
+    def sample(self):
+        """ Samples observations from several samplers. """
```

Comment:
Change to
"""
Samples observations from several samplers. 

:return: TODO
"""

AND

add return type hints, e.g. def foo() -> TODO: